src/resolveType.js

```diff
@@ -1,72 +1,36 @@
 const createProxy = require('react-proxy').default;
 
-let proxies = {};
-let warnedAboutTypes = {};
+let proxiesByID = {};
+let idsByType = new WeakMap();
 
-function setFlag(obj, key) {
-  try {
-    Object.defineProperty(obj, key, {
-      configurable: true,
-      enumerable: false,
-      value: true
-    });
-  } catch (err) {}
-}
+window.__REACT_HOT_LOADER__ = {
+  set(id, component) {
+    idsByType.set(component, id);
+  }
+};
 
 function resolveType(type) {
   // We only care about composite components
   if (!type || typeof type === 'string') {
     return type;
   }
 
-  // If the type is not tagged, return it as is.
-  if (
-    !Object.hasOwnProperty.call(type, '__source') ||
-    !type.__source ||
-    !type.__source.fileName ||
-    !type.__source.localName
-  ) {
-    setFlag(type, '__noSourceFound');
-    return type;
-  }
-
-  // Uniquely identifiy the component in code across reloads.
-  const source = type.__source;
-  const id = source.fileName + '#' + source.localName;
-
-  if (type.hasOwnProperty('__noSourceFound')) {
-    // This component didn't have a source last time, but now it has?
-    // This means createElement() was called during module definition.
-    // Bail out, or the component will be unmounted unexpectedly this time,
-    // as we'll return proxy but we returned the original class the last time.
-    // https://github.com/gaearon/react-hot-loader/issues/241
-    if (!warnedAboutTypes[id]) {
-      warnedAboutTypes[id] = true;
-      console.error(
-        `React Hot Loader: ${source.localName} from ${source.fileName} will not ` +
-        `hot reload correctly because it contains an imperative call like ` +
-        `ReactDOM.render() in the same file. Split ${source.localName} into a ` +
-        `separate file for hot reloading to work.`
-      );
-    }
+  var id = idsByType.get(type);
```

Comment:
The reason stale types were used is because Router cached old `type`s and called `createElement` with them. Now the only “entry point” to updating a proxy is an explicit `__REACT_HOT_LOADER__` call. So router can’t revive old types.